.github/workflows/publish.yml

```diff
@@ -0,0 +1,94 @@
+name: Docker publish
+
+on:
+  push:
+    branches:
+      - next
+      - qa
+      - main
+
+env:
+  # Use docker.io for Docker Hub if empty
+  REGISTRY: ghcr.io
+
+jobs:
+  build:
+
+    runs-on: ubuntu-latest
+    defaults:
+      run:
+        shell: bash -l {0}
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+      - uses: conda-incubator/setup-miniconda@v3
+        with:
+          auto-update-conda: true
+          channels: conda-forge,defaults
+          mamba-version: "*"
+          environment-file: environment.yml
+          cache-environment-key: ${{ runner.os }}-env-${{ hashFiles('**/environment.yml') }}
+          cache-downloads-key: ${{ runner.os }}-downloads-${{ hashFiles('**/environment.yml') }}
+      - name: Install requirements
+        run: |
+          # --quiet should turn off progress bars to make logs more readable
+          conda env create
+      - name: Get version
+        id: version
+        run: |
+          conda activate livedata
+          echo "version=$(versioningit)" >> $GITHUB_OUTPUT
+
+      - name: Create tag version
+        id: tag
+        run: |
+          conda activate livedata
+          echo "tag=$(versioningit).b$(date +'%Y%m%d%H%M')" >> $GITHUB_OUTPUT
+
+      - name: Create latest tag version
+        id: latest_tag
+        run: |
+          case ${{ github.ref }} in
+
+          refs/heads/next)
+            echo "latest_tag=latest-dev" >> $GITHUB_OUTPUT
+            ;;
+
+          refs/heads/qa)
+            echo "latest_tag=latest-test" >> $GITHUB_OUTPUT
+            ;;
+
+          refs/heads/main)
+            echo "latest_tag=latest-prod" >> $GITHUB_OUTPUT
+            ;;
+
+          *)
+            exit 1
+            ;;
+
+          esac
+
+      - name: Check tag names
+        run: |
+          echo ${{ steps.latest_tag.outputs.latest_tag }}
+          echo ${{ steps.tag.outputs.tag }}
+
+      - name: Log into registry ${{ env.REGISTRY }}
+        uses: docker/login-action@v3
+        with:
+          registry: ${{ env.REGISTRY }}
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Build and push docker image
+        uses: docker/build-push-action@v6
+        with:
+          context: .
+          file: Dockerfile
+          tags: |
+            ${{ env.REGISTRY }}/${{ github.repository }}/live_data_server:${{ steps.latest_tag.outputs.latest_tag }}
```

Comment:
I'm confused because I can't find any secret `REGISTRY` in the repo's settings :thinking: 
Are these images supposed to be stored in the container registry of repo `livedata-deploy`?

Also, can we follow the convention of [web_reflectivity](https://code.ornl.gov/reflectometry/web_reflectivity/container_registry) for storing images?

<img width="300" alt="capture" src="https://github.com/neutrons/live_data_server/assets/1136006/8050d903-e56d-4cea-be23-1322a3de0bad">

There's really no need to store more than one image per branch.

Reply:
> I'm confused because I can't find any secret REGISTRY in the repo's settings 🤔

Because it's not a secret. `REGISTRY` is defined in the `publish.yml`, line 12

> Are these images supposed to be stored in the container registry of repo livedata-deploy?

They will be stored in the github container registry, same as data_workflow.

There is a later story to update the deployment.

> Also, can we follow the convention of [web_reflectivity](https://code.ornl.gov/reflectometry/web_reflectivity/container_registry) for storing images?

I was following the convention of [data_workflow](https://github.com/neutrons/data_workflow/blob/next/.github/workflows/publish.yml)

> There's really no need to store more than one image per branch.

What happens if you want to role back the deployed version, you will need to rebuild the docker image?

I can change this all to be more consistent with web_reflectivity instead of data_workflow if you really want but I though it made sense to make it the same as data_workflow since these are both on github and not code.ornl.gov?

Reply:
[Here](https://github.com/neutrons/data_workflow/pkgs/container/data_workflow%2Fwebmon) are the webmon images so you can see how they work.

Reply:
[Here](https://code.ornl.gov/sns-hfir-scse/infrastructure/web-monitor/-/issues/180) is the issue John wrote for the docker image naming convention for webmon, is there something similar for webref?

Reply:
After discussing with @backmari, we agreed on publishing only the following tags:
- `latest-dev`, `latest-test`, `lated-prod`
- production tags (but not release candidates)
Thus, line `echo "tag=$(versioningit).b$(date +'%Y%m%d%H%M')" >> $GITHUB_OUTPUT` should run just for branch `main`  and I think it should  be `echo "tag=$(versioningit)" >> $GITHUB_OUTPUT`  because we only update `main` when we create a production version.

We'll enforce this convention in webmon because a look to the container registry shows it's full of useless images:
<img width="866" alt="capture" src="https://github.com/neutrons/live_data_server/assets/1136006/18da43e0-d250-4ea7-b504-07e4eb0e9722">